sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+the desired state described by its inputs and its actual state. If the resource does not
+exist (i.e. there is no last refereshed state for its URN), the engine calls the
+provider's [`Create`](#create) method, which returns the ID and state of the created resource. If the
+resource does exist, the action taken depends on the differences (if any) between the
+desired and actual state of the resource.
+
+If the resource does exist, the engine calls the provider's [`Diff`](#diff) method with the
+inputs returned from [`Check`](#check), the resource's ID, and the resource's last refreshed state.
+If the result of the call indicates that there is no difference between the desired and
+actual state, no operation is necessary. Otherwise, the resource is either updated (if
+[`Diff`](#diff) does not indicate that the resource must be replaced) or replaced (if [`Diff`](#diff) does
+indicate that the resource must be replaced).
+
+To update a resource, the engine calls the provider's [`Update`](#update) method with the inputs
+returned from [`Check`](#check), the resource's ID, and its last refreshed state. [`Update`](#update) returns
+the new state of the resource. The resource's ID may not be changed by a call to [`Update`](#update).
+
+To replace a resource, the engine first calls [`Check`](#check) with an empty set of prior inputs
+and the inputs supplied with the resource's registration. If [`Check`](#check) fails, the resource
+is not replaced. Otherwise, the inputs returned by this call to [`Check`](#check) will be used to
+create the replacement resource. Next, the engine inspects the resource options supplied
+with the resource's registration and result of the call to [`Diff`](#diff) to determine whether
+the replacement can be created before the original resource is deleted. This order of
+operations is preferred when possible to avoid downtime due to the lag between the
+deletion of the current resource and creation of its replacement. If the replacement may
+be created before the original is deleted, the engine calls the provider's [`Create`](#create) method
+with the re-checked inputs, then later calls [`Delete`](#delete) with the resource's ID and original
+state. If the resource must be deleted before its replacement can be created, the engine
+first deletes the transitive closure of resource that depend on the resource being
+replaced. Once these deletes have completed, the engine deletes the original resource by
+calling the provider's [`Delete`](#delete) method with the resource's ID and original state. Finally,
+the engine creates the replacement resource by calling [`Create`](#create) with the re-checked
+inputs.
+
+If a managed resource registered by a Pulumi program is not re-registered by the next
+successful execution of a Pulumi progam in the resource's stack, the engine deletes the
+resource by calling the resource's provider's [`Delete`](#delete) method with the resource's ID and
+last refereshed state.
```

Comment:
There's also another case where pending deletes can get processed at the beginning of an update - right?